crate2nix/src/prefetch.rs

```diff
@@ -93,22 +114,51 @@ pub fn nix_prefetch_from_crate_io(
             crate_derivation.crate_name, crate_derivation.version
         ),
     ];
-    let output = Command::new(cmd)
-        .args(&args)
-        .output()
-        .map_err(|e| format_err!("While spawning '{} {}': {}", cmd, args.join(" "), e))?;
+    get_command_output(cmd, &args)
+}
 
-    if !output.status.success() {
-        std::io::stdout().write_all(&output.stdout)?;
-        std::io::stderr().write_all(&output.stderr)?;
-        bail!(
-            "{}\n=> exited with: {}",
-            "nix-prefetch-url",
-            output.status.code().unwrap_or(-1)
-        );
-    }
+/// A struct used to contain the output returned by `nix-prefetch-git`.
+///
+/// Additional fields are available (e.g., `name`), but we only call `nix-prefetch-git` to obtain
+/// the nix sha256 for use in calls to `pkgs.fetchgit` in generated `Cargo.nix` files so there's no
+/// reason to declare the fields here until they are needed.
+#[derive(Deserialize)]
+struct NixPrefetchGitInfo {
+    sha256: String,
+}
 
-    Ok(String::from_utf8(output.stdout)
-        .map(|s| s.trim().to_string())
-        .map_err(|_e| format_err!("output of '{} {}' is not UTF8!", cmd, args.join(" ")))?)
+fn nix_prefetch_from_git(
+    crate_derivation: &CrateDerivation,
+    idx: usize,
+    num_packages: usize,
+) -> Result<String, Error> {
+    if let ResolvedSource::Git {
+        url, rev, r#ref, ..
+    } = &crate_derivation.source
+    {
+        eprintln!("Prefetching {:>4}/{}: {}", idx, num_packages, url);
+        let cmd = "nix-prefetch-git";
+        let mut args = vec![
+            "--url",
+            url.as_str(),
+            "--fetch-submodules",
+            "--rev",
+            rev,
+        ];
+
+        // TODO: --branch-name isn't documented in nix-prefetch-git --help
+        // TODO: Consider the case when ref *isn't* a branch. You have to pass
+        // that to `--rev` instead. This seems like limitation of nix-prefetch-git.
+        if let Some(r#ref) = r#ref {
```

Comment:
Could this produce a wrong hash then somehow?

Could we fall back and to not returning the hash and caching that prefetch doesn't work?

Reply:
I don't think it can return an invalid hash, but it may cause the prefetch command to fail if something other than a branch is used for the `r#ref`. Currently that can't be the case IIUC, but I suspect we want to allow tags in the future.